zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
+		depth.copyTo(destination, mask);
+
+		cv::normalize(destination, destination, 0, 255, cv::NORM_MINMAX);
+		destination.convertTo(destination, CV_8UC1);
+		cv::imshow("Masked", destination);
+		cv::waitKey(0);
+	}
+
+	// copy matrix data to a vector
+	std::vector<float> vec;
+	masked = masked.reshape(0, 1);
+	masked.copyTo(vec);
+	// remove 999 values from the mask
+	vec.erase(std::remove(vec.begin(), vec.end(), 999), vec.end());
+	// and 0 values (no depth)
+	vec.erase(std::remove(vec.begin(), vec.end(), 0), vec.end());
+	// sort vector
+	std::sort(vec.begin(), vec.end());
+
+	return (vec.size() != 0 ? vec[0] : -1); // return lowest value. Ideally, we'd want to throw out the bottom
+	// <some number>%, but that messes up depth calculation for things like spheres.
+	// If there are no values, return -1.
+
+	/* Test results (adaptive thresholding):
+	[ INFO] [1635948213.029497061]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal_8.jpg
+	[ INFO] [1635948213.689598629]: Calculated depth is 113.001 ✅
+	[ INFO] [1635948229.253951900]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal_behind_power_cell.png
+	[ INFO] [1635948229.870174338]: Calculated depth is 138.001 ✅
+	[ INFO] [1635948236.848476178]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal.png
+	[ INFO] [1635948237.445102845]: Calculated depth is 113.001 ✅
+	[ INFO] [1635948248.428153064]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_loading_bay.png
+	[ INFO] [1635948249.227361594]: Calculated depth is 141.001 ✅
+	[ INFO] [1635948255.684437760]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_low_goal.png
+	[ INFO] [1635948256.286312490]: Calculated depth is 92.001 ✅
+	[ INFO] [1635948266.234706198]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_power_cell.png
+	[ INFO] [1635948266.316279430]: Calculated depth is 0.001 ✅
+	[ INFO] [1635948273.702984310]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_behind_power_cell.png
+	[ INFO] [1635948275.086796499]: Calculated depth is 0.001 **non-cropped version** ❌
+	[ INFO] [1635948302.568242461]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_with_power_cell_in_front.jpg
+	[ INFO] [1635948302.598558216]: Calculated depth is 0.001 ❌
+	[ INFO] [1635948312.950812216]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/test_weirdness.jpg
+	[ INFO] [1635948312.980837687]: Calculated depth is 73.001 ✅
+	*/
+}
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using k-means
+double kMeansDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect, bool debug = false, int maximumK = 3, float tolerance = 1e-3)
```

Comment:
Same nitpicking here on reordering the arguments

Reply:
Also, if you make maximumK unsigned, you can get rid of some casting later in the code. Kinda makes semantic sense too - a negative minimum K would be a bad thing.

Reply:
Ok, `maximumK` is now an `unsigned int` -- thanks!

Reply:
edit: as I read more of this, I saw that you said I should use `size_t`, so `maximumK` and `k` are `size_t` variables now